### PR TITLE
Fix GetTempPath

### PR DESCRIPTION
**Pull request by JeremyKuhne**
GetTempPath and GetTempFileName weren't updated to handle long paths.
Update to use API properly and stop using StringBuilder. Also tweak
Normalize to allow utilizing an existing Span as input.

cc: @jkotas, @danmosemsft, @Anipik, @pjanotti

### CODE REVIEW

File: src/mscorlib/shared/Interop/Windows/Kernel32/Interop.GetFullPathNameW.cs [hunk 1]
```diff
@@ -23,7 +23,7 @@ internal static unsafe uint GetFullPathNameW(string path, uint numBufferChars, r
                 return GetFullPathNameW(path, numBufferChars, pBuffer, mustBeZero);
         }
 #else
-        internal static extern uint GetFullPathNameW(string path, uint numBufferChars, ref char buffer, IntPtr mustBeZero);
```
Comment (jkotas) [inline src/mscorlib/shared/Interop/Windows/Kernel32/Interop.GetFullPathNameW.cs]: Could you please update the PROJECTN workaround above as well?

File: src/mscorlib/shared/Interop/Windows/Kernel32/Interop.GetFullPathNameW.cs [hunk 2]
```diff
@@ -23,7 +23,7 @@ internal static unsafe uint GetFullPathNameW(string path, uint numBufferChars, r
...
+        internal static extern uint GetFullPathNameW(ref char path, uint numBufferChars, ref char buffer, IntPtr mustBeZero);
```
Comment (jkotas) [inline src/mscorlib/shared/Interop/Windows/Kernel32/Interop.GetFullPathNameW.cs]: Nit: It may be nice to drop the W suffix from these names while you are on it, and have it in the `EntryPoint = "GetFullPathNameW"` only. It is how all other files except these few are structured.

File: src/mscorlib/shared/System/IO/PathHelper.Windows.cs [hunk 3]
```diff
@@ -42,14 +42,31 @@ internal static string Normalize(string path)
...
+
+        internal static void GetFullPathName(ReadOnlySpan<char> path, ref ValueStringBuilder builder)
         {
             // If the string starts with an extended prefix we would need to remove it from the path before we call GetFullPathName as
             // it doesn't root extended paths correctly. We don't currently resolve extended paths, so we'll just assert here.
             Debug.Assert(PathInternal.IsPartiallyQualified(path) || !PathInternal.IsExtended(path));
 
             uint result = 0;
-            while ((result = Interop.Kernel32.GetFullPathNameW(path, (uint)builder.Capacity, ref builder.GetPinnableReference(), IntPtr.Zero)) > builder.Capacity)
+            while ((result = Interop.Kernel32.GetFullPathNameW(ref MemoryMarshal.GetReference(path), (uint)builder.Capacity, ref builder.GetPinnableReference(), IntPtr.Zero)) > builder.Capacity)
```
Comment (jkotas) [inline src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: What guarantees that the path is zero terminated?

File: src/mscorlib/shared/System/IO/PathHelper.Windows.cs [hunk 4]
```diff
@@ -42,14 +42,31 @@ internal static string Normalize(string path)
...
+
+            // Get the full path
+            GetFullPathName(path, ref builder);
+
+            string result = builder.AsSpan().Contains('~')
+                ? TryExpandShortFileName(ref builder, originalPath: null)
+                : builder.ToString();
+
+            // Clear the buffer
+            builder.Dispose();
```
Comment (danmosemsft) [inline src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: any reason you don't use a using statement

Comment (danmosemsft) [reply src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: same elsewhere for ValueStringBuilder.

File: src/mscorlib/shared/System/IO/PathHelper.Windows.cs (hunk 3 above)
Comment (jkotas) [reply src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: I think this may be going overboard with trying to avoid allocations. Bad tradeoff between simplicity/maintainability and performance.

File: src/mscorlib/shared/System/IO/Path.Windows.cs [hunk 5]
```diff
@@ -128,24 +128,60 @@ public static string GetFullPath(string path, string basePath)
 
         public static string GetTempPath()
         {
-            StringBuilder sb = StringBuilderCache.Acquire(Interop.Kernel32.MAX_PATH);
-            uint r = Interop.Kernel32.GetTempPathW(Interop.Kernel32.MAX_PATH, sb);
-            if (r == 0)
+            Span<char> initialBuffer = stackalloc char[PathInternal.MaxShortPath];
+            ValueStringBuilder builder = new ValueStringBuilder(initialBuffer);
```
Comment (danmosemsft) [inline src/mscorlib/shared/System/IO/Path.Windows.cs]: Nit, you could use var.

File: src/mscorlib/shared/System/IO/PathHelper.Windows.cs [hunk 6]
```diff
@@ -220,7 +237,7 @@ internal static string TryExpandShortFileName(ref ValueStringBuilder outputBuild
             // Strip out any added characters at the front of the string
             ReadOnlySpan<char> output = builderToUse.AsSpan(rootDifference);
 
-            string returnValue = output.EqualsOrdinal(originalPath.AsSpan())
+            string returnValue = !string.IsNullOrEmpty(originalPath) && output.EqualsOrdinal(originalPath.AsSpan())
```
Comment (jkotas) [inline src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: You do not need the check for Empty string. Maybe this can be just:

`string returnValue = ((originalPath != null) && output.EqualsOrdinal(originalPath)) ? originalPath : new string(output);`

File: src/mscorlib/shared/System/IO/Path.Windows.cs [hunk 7]
```diff
@@ -128,24 +128,60 @@ public static string GetFullPath(string path, string basePath)
...
-            StringBuilder sb = StringBuilderCache.Acquire(Interop.Kernel32.MAX_PATH);
-            uint r = Interop.Kernel32.GetTempFileNameW(path, "tmp", 0, sb);
-            if (r == 0)
+            tempPathBuilder.Dispose();
+
+            if (result == 0)
                 throw Win32Marshal.GetExceptionForLastWin32Error();
-            return StringBuilderCache.GetStringAndRelease(sb);
+
+            builder.Length = (int)result;
```
Comment (jkotas) [inline src/mscorlib/shared/System/IO/Path.Windows.cs]: Nit: `builder.Length = (int)result;` belongs logically to the previous block. Add empty line after, and remove the empty line before?

File: src/mscorlib/shared/System/IO/PathHelper.Windows.cs (hunk 3 above)
Comment (JeremyKuhne) [reply src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: The APIs we call null terminate. I can add an unsafe assert here.

File: src/mscorlib/shared/System/IO/PathHelper.Windows.cs (hunk 4 above)
Comment (JeremyKuhne) [reply src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: Can't. It is a ref struct. :/ Per prior feedback we're ok with not returning buffers to the pool on the error case.

File: src/mscorlib/shared/System/IO/PathHelper.Windows.cs (hunk 6 above)
Comment (JeremyKuhne) [reply src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: Sure, just habit. :)

**Comment by jkotas** [comment]
Can Windows actually return temp path that is longer than `MAX_PATH` today? The MSDN docs say that it cannot return anything longer https://msdn.microsoft.com/en-us/library/windows/desktop/aa364992(v=vs.85).aspx .

File: src/mscorlib/shared/Interop/Windows/Kernel32/Interop.GetFullPathNameW.cs (hunk 2 above)
Comment (JeremyKuhne) [reply src/mscorlib/shared/Interop/Windows/Kernel32/Interop.GetFullPathNameW.cs]: I've been doing it this way to save a string. Over the whole space it isn't huge, but also isn't nothing.

File: src/mscorlib/shared/Interop/Windows/Kernel32/Interop.GetFullPathNameW.cs (hunk 1 above)
Comment (JeremyKuhne) [reply src/mscorlib/shared/Interop/Windows/Kernel32/Interop.GetFullPathNameW.cs]: Sure, fixing

File: src/mscorlib/shared/System/IO/PathHelper.Windows.cs (hunk 4 above)
Comment (danmosemsft) [reply src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: Ah of course, cannot assign to interface type.
https://docs.microsoft.com/en-us/dotnet/csharp/reference-semantics-with-value-types#ref-struct-type

File: src/mscorlib/shared/Interop/Windows/Kernel32/Interop.GetFullPathNameW.cs (hunk 2 above)
Comment (jkotas) [reply src/mscorlib/shared/Interop/Windows/Kernel32/Interop.GetFullPathNameW.cs]: I do not think it makes sense to worry about manual micro savings like this. I think consistency is more important. 

If we were worried about savings like this - we would be compressing all method names to 3 letters.

Reducing space occupied by method names is a job for ILLink or AOT toolchain.

Comment (JeremyKuhne) [reply src/mscorlib/shared/Interop/Windows/Kernel32/Interop.GetFullPathNameW.cs]: > I think consistency is more important.

But isn't it consistent to actually match the entry point name? I like the explicitness of it and searching gets you to the right doc page so I'm not sure what it buys to not use the actual name.

**Comment by JeremyKuhne** [comment]
> Can Windows actually return temp path that is longer than MAX_PATH today?

Yes, although having one isn't likely to work- particularly as we don't support it. :)

File: src/mscorlib/shared/System/IO/Path.Windows.cs (hunk 7 above)
Comment (JeremyKuhne) [reply src/mscorlib/shared/System/IO/Path.Windows.cs]: Tweaked

File: src/mscorlib/shared/Interop/Windows/Kernel32/Interop.GetFullPathNameW.cs (hunk 2 above)
Comment (jkotas) [reply src/mscorlib/shared/Interop/Windows/Kernel32/Interop.GetFullPathNameW.cs]: It depends on the convention. Majority of Windows interop in CoreCLR and CoreFX is on the specify W suffix in entrypoint only plan, so it is a de-facto convention.

The W suffix is more explicit, but also more ugly. I do not think there is obvious answer which one is right. I think that there is value in being consistent across the project.

File: src/mscorlib/shared/System/IO/PathHelper.Windows.cs (hunk 3 above)
Comment (jkotas) [reply src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: You are right that it happens to work today because of there is zero termination done 3 hops away, but it very hard to follow.

The regular contract for Span<char> is to not be zero terminated. There should be comment next to every Span local and argument that is assumed to be zero terminated.

File: src/mscorlib/shared/Interop/Windows/Kernel32/Interop.GetFullPathNameW.cs (hunk 2 above)
Comment (JeremyKuhne) [reply src/mscorlib/shared/Interop/Windows/Kernel32/Interop.GetFullPathNameW.cs]: I also prefer using the actual parameter names which fits into the ugly bucket as well I suppose. To me being super explicit is a sign we're paying attention. :) For parameter names there is a documentation aspect to it as well- if your signature param types don't match the Hungarian it raises red flags.

File: src/mscorlib/shared/System/IO/PathHelper.Windows.cs (hunk 3 above)
Comment (JeremyKuhne) [reply src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: I'll make sure the comment is in there. It raises an interesting question about interop and ValueStringBuilder/string/Span. We don't have a great way to represent a null terminated Span. Should we have ToSpanWithNull()?

Comment (jkotas) [reply src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: ToSpanWithNull would be a way to get zero a null terminated span, but it is not a way to represent a zero terminated span.

I agree that it would be nice to have a set of conventions on how to represent zero terminated spans.

We are doing `Append('\0')` and `GetPinnableReference` today that I do not think is too bad to get a zero terminated string. Not sure whether `ToSpanWithNull` method would be an improvement.

File: src/mscorlib/shared/Interop/Windows/Kernel32/Interop.GetFullPathNameW.cs (hunk 2 above)
Comment (jkotas) [reply src/mscorlib/shared/Interop/Windows/Kernel32/Interop.GetFullPathNameW.cs]: I think using actual parameter names is fine. It is the de-facto convention for CoreCLR and CoreFX interop. I have done a random sampling and I believe that majority of the CoreCLR and CoreFX interop code is using the actual argument names already.

File: src/mscorlib/shared/System/IO/Path.Windows.cs (hunk 5 above)
Comment (ahsonkhan) [reply src/mscorlib/shared/System/IO/Path.Windows.cs]: I believe the guideline is that we should use `var` whenever the type can be statically inferred from the right hand side. In most cases, that means, wherever the `new` keyword is used, we likely have a candidate for use of `var` (or when there is explicit casting). I would be curious to see in what instance you got feedback contrary to that.

File: src/mscorlib/shared/System/IO/PathHelper.Windows.cs (hunk 3 above)
Comment (JeremyKuhne) [reply src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: Discussed this some with @KrzysztofCwalina and opened https://github.com/dotnet/corefx/issues/28040 for getting the terminator from the string. I'm updating the pull here to add the assert in such a manner.

File: src/mscorlib/shared/System/IO/Path.Windows.cs (hunk 5 above)
Comment (JeremyKuhne) [reply src/mscorlib/shared/System/IO/Path.Windows.cs]: The guideline doesn't mandate it. https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md

Additionally I don't think the guideline is clear enough. It borders on gratuitous in some cases (`var values = new int[30]`, `var info = new FileInfo(path)`). This particular instance is a little grey, so I'm ok with var, but I don't think it is necessary. If I make another update I'll change it.

File: src/mscorlib/shared/System/IO/PathHelper.Windows.cs [hunk 8]
```diff
@@ -42,14 +42,33 @@ internal static string Normalize(string path)
             return result;
         }
 
-        private static void GetFullPathName(string path, ref ValueStringBuilder builder)
+        internal static string Normalize(ReadOnlySpan<char> path)
```
Comment (jkotas) [inline src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: Would it be better for Normalize to take the ValueStringBuilder as argument (non-zero terminated) and zero terminate it before the PInvoke?

It would also address my concern of zero-terminated `Span<char>`s  leaking through the system.

Comment (JeremyKuhne) [reply src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: I'll give it a look and get back to you shortly.

File: src/mscorlib/shared/System/Text/ValueStringBuilder.cs [hunk 9]
```diff
@@ -58,7 +70,20 @@ public override string ToString()
             return s;
         }
 
-        public ReadOnlySpan<char> AsSpan() => _chars.Slice(0, _pos);
+        /// <summary>
+        /// Returns a span around the contents of the builder.
+        /// </summary>
+        /// <param name="terminate">Ensures that the builder has a null char after <see cref="Length"/></param>
+        public ReadOnlySpan<char> AsSpan(bool terminate = false)
```
Comment (jkotas) [inline src/mscorlib/shared/System/Text/ValueStringBuilder.cs]: Is this going to make the method go over the inlining threshold?

Comment (JeremyKuhne) [reply src/mscorlib/shared/System/Text/ValueStringBuilder.cs]: I don't know the full rules. I can split/attribute?

Comment (danmosemsft) [reply src/mscorlib/shared/System/Text/ValueStringBuilder.cs]: Perfview can show you whether something was innings and if not why not.

File: src/mscorlib/shared/System/IO/PathHelper.Windows.cs [hunk 10]
```diff
@@ -26,30 +26,57 @@ internal class PathHelper
...
-            GetFullPathName(path, ref builder);
+            GetFullPathName(path.AsSpan(), ref builder);
 
             // If we have the exact same string we were passed in, don't allocate another string.
             // TryExpandShortName does this input identity check.
-            string result = builder.AsSpan().Contains('~')
+            string result = builder.AsSpan().IndexOf('~') >= 0
                 ? TryExpandShortFileName(ref builder, originalPath: path)
-                : builder.AsSpan().EqualsOrdinal(path.AsSpan()) ? path : builder.ToString();
+                : builder.AsSpan().Equals(path.AsSpan(), StringComparison.Ordinal) ? path : builder.ToString();
```
Comment (ahsonkhan) [inline src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: Why is this change from EqualsOrdinal to Equals necessary?
We end up calling the EqualsOrdinal helper internally, anyway https://github.com/dotnet/coreclr/blob/7ac6d5cd8db4033a4d7a4bfec34e72701b13433b/src/mscorlib/shared/System/MemoryExtensions.Fast.cs#L56

Comment (JeremyKuhne) [reply src/mscorlib/shared/System/IO/PathHelper.Windows.cs]: Because I'm using the source file in https://github.com/dotnet/corefx/pull/28089.  (And will use it elsewhere.)
